api/admin/controller/__init__.py

```diff
@@ -372,6 +374,63 @@ def __call__(self, collection, book, path=None):
         response.set_cookie("csrf_token", csrf_token, httponly=True)
         return response
 
+class TimestampsController(AdminCirculationManagerController):
+    """Returns a dict: each key is a type of service (script, monitor, or coverage provider);
+    each value is a nested dict in which timestamps are organized by service name and then by collection ID."""
+
+    def diagnostics(self):
+        self.require_system_admin()
+        timestamps = self._db.query(Timestamp).order_by(Timestamp.start)
+        if isinstance(timestamps, ProblemDetail):
+            return timestamps
+        sorted = self._sort_by_type(timestamps)
+        for type in sorted.keys():
+            for service in sorted[type]:
+                by_collection = self._sort_by_collection(sorted[type][service])
+                sorted[type][service] = by_collection
+
+        return sorted
+
+    def _sort_by_type(self, timestamps):
+        """Takes a list of Timestamp objects.  Returns a dict: each key is a type of service
+        (script, monitor, or coverage provider); each value is a dict in which the keys are the names
+        of services and the values are lists of timestamps."""
+
+        result = {}
+        for type in ["script", "monitor", "coverage_provider"]:
+            filtered = [self._extract_info(ts) for ts in timestamps if ts.service_type == type]
+            result[type] = self._sort_by_service(filtered)
+        return result
+
+    def _sort_by_service(self, timestamps):
+        """Returns a dict: each key is the name of a service; each value is a list of timestamps."""
+
+        result = {}
+        for timestamp in timestamps:
+            result.setdefault(timestamp.get("service"), []).append(timestamp)
+        return result
+
+    def _sort_by_collection(self, timestamps):
+        """Takes a list of timestamps; turns it into a dict in which each key is a
+        collection ID and each value is a list of the timestamps associated with that collection."""
+
+        result = {}
+        for timestamp in timestamps:
+            result.setdefault((timestamp.get("collection_id") or -1), []).append(timestamp)
+        return result
+
+    def _extract_info(self, timestamp):
+        """Takes a Timestamp object and returns a dict"""
+
+        return dict(
+            id=timestamp.id,
+            start=timestamp.start,
+            duration=(str(timestamp.finish - timestamp.start)),
```

Comment:
I don't think this will happen in real life, but it could theoretically. You should be prepared for `start` or `finish` to be null, which will crash this code. Before making the dictionary, calculate `duration` and set it to null if either `start` or `finish` is null.

Also, you can convert this into a number of seconds with `(timestamp.finish - timestamp.start).total_seconds()`. Then on the client side you can format the number of seconds however you like.